### ISSUE TITLE
Warn on unhandled error.

### ISSUE THREAD

**Issue by PaperMountainStudio** [mgechev/revive#120]
Sometimes I forget to handle an error. Go linters usually don't warn me. At least those I know. I don't think I am the only one who is bothered by this. By implementing this everyone's code will greatly improve.
For example:

Bad code:
```go
os.Chdir("..")
```
But if we look into the docs, we can see that `os.Chdir` returns error.
Good code:
```go
err = os.Chdir("..")
if err != nil {
    // something
}
```
Actually, we don't need to check for the `if err != nil` part. It is handled by go compiler. We just need to check if err was assigned.

**Solution I would like**
Error like this:
`os.Chdir error not handled`

Thanks.

**Comment by chavacava** [comment]
Hi @PaperMountainStudio, if I understand well you want to be warned in cases like:
```go
fmt.Fprintf(w, "...")
```

or if 
```go
func foo() (int, err) { ... }
```
you call it 
```go
r,_ := foo()
foo()
```

**Comment by PaperMountainStudio** [comment]
Hi @chavacava  I don't know. In your example with assigning error to _, it seems fine. You chose to ignore the error.
But your second case is what I meant.
```go
_ = foo() // fine?
foo() // wrong
```

**Comment by chavacava** [comment]
I get it. I like the rule.
I see some problems to implement it. The rule needs to known the signature of functions in order check for unhanded errors. Currently, a rule applies to a file, then it has only information about functions declared in the file.
We can try to provide a _dictionary_ of functions but this approach requires the maintenance of the dictionary to have decent confidence on the results.
We can also try to build the dictionary on-the-fly by parsing, at least, the local copy of the standard lib source code, but this may require some kind of config and, most important, lot of processing before applying the rule.
Do you have other ideas?

**Comment by mgechev** [comment]
How about using the type checker? We can get the type of the result of each function invocation, and from there we can decide if we want to throw or not. The `TypeCheck` is per package, which will let us resolve non-local information.

**Comment by chavacava** [comment]
@mgechev Yes the type checker is an option but the rule will cope only with functions declared in the pkg. In the initial example of @PaperMountainStudio, the rule needs to handle a standard library function.

We can develop a first version of the rule based on the type checker and extend it later to cope with _global_ cases

**Comment by chavacava** [comment]
Good news, the type checking provides information of standard lib funcs.

For the following code:
```go
package fixtures

import "fmt"

func unhandledError1(a int) (int, error) {
	return a, nil
}

func unhandledError2() error {
	_, err := unhandledError1(1)
	unhandledError1(1)
	fmt.Fprintf(nil, "")
	return err
}
```

I get the following info
```
Func call of type func(a int) (int, error)
Func call of type func(a int) (int, error)
Func call of type func(w io.Writer, format string, a ...interface{}) (n int, err error)
```

**Comment by chavacava** [comment]
@PaperMountainStudio 
You can checkout and test the first working version of the rule from this PR #122 
Feedback is wellcome

**Comment by PaperMountainStudio** [comment]
@chavacava This rule wasn't enabled by default, so I had to manually add `[rule.unhandled-error]` to the end of defaults.toml and call revive with this config file.
Otherwise, it is perfect. I found more unhandled errors in my code than I imagined. :smile:
